### PR TITLE
[Example] Hypergraph attention

### LINKED ISSUE #4871: [Sparse] Hypergraph Convolution and Hypergraph Attention
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
Implements hypergraph attention in https://arxiv.org/pdf/1901.08150.pdf.
The hypergraph convolution there is identical to the HGNN example in #4917 .
Closes #4871 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/sparse/hypergraphatt.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
+"""
+Hypergraph Convolution and Hypergraph Attention
+(https://arxiv.org/pdf/1901.08150.pdf).
+"""
+import dgl
+import dgl.data
```
Comment (mufeili) [inline examples/sparse/hypergraphatt.py:6]: No need to import `dgl.data`

**Comment by mufeili** [comment]
Any reference accuracy number?

File: examples/sparse/hypergraphatt.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torchmetrics.functional import accuracy
+import tqdm
+
+# Question:
+# The paper proposed two models, but HypergraphConv is essentially
+# what is in the HGNN paper (identical).  So should I rewrite HGNN
+# in the same fashion as this file?
```
Comment (mufeili) [inline examples/sparse/hypergraphatt.py:17]: I think it's better to be consistent.

File: examples/sparse/hypergraphatt.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+def hypergraph_laplacian(H):
+    ###########################################################
+    # (HIGHLIGHT) Compute the Laplacian with Sparse Matrix API
+    ###########################################################
+    d_V = H.sum(1)  # node degree
+    d_E = H.sum(0)  # edge degree
```
Comment (mufeili) [inline examples/sparse/hypergraphatt.py:23]: Maybe just say `d_V = d_E = H.sum(0)`?

File: examples/sparse/hypergraphatt.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    return laplacian
+
+class HypergraphAttention(nn.Module):
+    """Hypergraph Attention module.
+
+    As per Section 3.3 in the paper
+    `Hypergraph Convolution and Hypergraph Attention
+    <https://arxiv.org/pdf/1901.08150.pdf>`_, this module only works when
+    the number of hyperedges equal the number of nodes.
```
Comment (mufeili) [inline examples/sparse/hypergraphatt.py:38]:
```suggestion
    the number of hyperedges equals to the number of nodes.
```

File: examples/sparse/hypergraphatt.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        self.a = nn.Linear(2 * out_size, 1)
+
+    def forward(self, H, X, X_edges):
+        Z = self.P(X)
+        Z_edges = self.P(X_edges)
+        sim = self.a(torch.cat([Z[H.row], Z_edges[H.col]], 1))
+        sim = F.leaky_relu(sim, 0.2).squeeze(1)
+        # Reassign the hypergraph with new weights
+        H_att = dglsp.create_from_coo(H.row, H.col, sim, shape=H.shape)
+        H_att = H_att.softmax()
```
Comment (mufeili) [inline examples/sparse/hypergraphatt.py:53]: dim?

**Review by mufeili — APPROVED** [review]
Some minor comments

File: examples/sparse/hypergraphatt.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    ###########################################################
+    # (HIGHLIGHT) Compute the Laplacian with Sparse Matrix API
+    ###########################################################
+    d_V = H.sum(1)  # node degree
+    d_E = H.sum(0)  # edge degree
+    n_edges = d_E.shape[0]
+    D_V_invsqrt = dglsp.diag(d_V ** -0.5)  # D_V ** (-1/2)
+    D_E_inv = dglsp.diag(d_E ** -1)  # D_E ** (-1)
+    W = dglsp.identity((n_edges, n_edges))
+    laplacian = D_V_invsqrt @ H @ W @ D_E_inv @ H.T @ D_V_invsqrt
```
Comment (frozenbugs) [inline examples/sparse/hypergraphatt.py:28]: nit:
```
# Laplacian.
return D_V_invsqrt @ H @ W @ D_E_inv @ H.T @ D_V_invsqrt
```

File: examples/sparse/hypergraphatt.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+        self.P = nn.Linear(in_size, out_size)
+        self.a = nn.Linear(2 * out_size, 1)
+
+    def forward(self, H, X, X_edges):
+        Z = self.P(X)
+        Z_edges = self.P(X_edges)
+        sim = self.a(torch.cat([Z[H.row], Z_edges[H.col]], 1))
+        sim = F.leaky_relu(sim, 0.2).squeeze(1)
+        # Reassign the hypergraph with new weights
```
[the remark this answers is not in the document]
Comment (frozenbugs) [reply examples/sparse/hypergraphatt.py:51]: and end with dot.

**Review by frozenbugs — APPROVED** [review]

**Comment by BarclayII** [comment]
Accuracy is 80.1.

File: examples/sparse/hypergraphatt.py (hunk 2 above)
Comment (BarclayII) [reply examples/sparse/hypergraphatt.py:17]: I can do it in later PRs.

File: examples/sparse/hypergraphatt.py (hunk 4 above)
Comment (BarclayII) [reply examples/sparse/hypergraphatt.py:38]: This description is actually outdated as I included the hyperedge representations as input as well (same as PyG).  So I'll remove it.

File: examples/sparse/hypergraphatt.py (hunk 3 above)
Comment (BarclayII) [reply examples/sparse/hypergraphatt.py:23]: I wrote the implementation as if the number of hyperedges does not equal the number of nodes so that people can adapt the code to other hypergraph datasets more easily.
